File: RL_Maggie/Code/02_trial_etl_fibro.R
## 02_trial_etl_fibro.R --------------------------------------------------
get_trial_data_healthy_fibro <- function(data_dir = LOCAL_DATA_DIR) {
  files <- list.files(data_dir, "^sub_.*\\.csv$", full.names = TRUE)
  
  df <- tibble();   participants_with_7 <- character()
  
  for (f in files) {
    if (file.info(f)$size == 0) { message("Empty file: ", f); next }
    
    sub_df <- read.csv(f) %>% filter(!is.na(trial_num))
    if (nrow(sub_df) == 0)   { message("No trials in: ", f); next }
    
    sub_df <- sub_df %>% 
      select(trial_num, participant, pair_type, choice_a, reward,
             correct, high_prob_image_file, low_prob_image_file) %>% 
      mutate(block = rep(1:ceiling(n()/20), each = 20)[1:n()])
    
    cur_p <- sub_df %>% group_by(participant) %>% 
      summarise(total_blocks = max(block), .groups = "drop") %>% 
      filter(total_blocks == 7) %>% pull(participant)
    participants_with_7 <- c(participants_with_7, cur_p)
    
    if (max(sub_df$block) == 7) {
      sub_df <- sub_df %>% filter(block != 1) %>% 
        mutate(block = block - 1, trial_num = trial_num - 20)
    }
    df <- bind_rows(df, sub_df)
  }
  
  df <- df %>% mutate(correct = if_else(correct == "correct", 1, 0))
  participants_with_7 <- unique(participants_with_7)
  
  ## Non-learners ---------------------------------------------------------
  non_learners <- df %>% 
    filter(participant %in% participants_with_7, block == 3) %>% 
    group_by(participant, pair_type) %>% 
    summarise(pct = mean(correct) * 100, .groups = "drop") %>% 
    pivot_wider(names_from = pair_type, values_from = pct) %>% 
    filter(reversed   < 70 | `non-reversed` < 70) %>% pull(participant)
  
  ## Group labels ---------------------------------------------------------
  p_nums <- as.numeric(str_remove(unique(df$participant), "sub_"))
  healthy <- unique(df$participant)[p_nums >= 900 & p_nums <= 1499]
  fibro   <- setdiff(unique(df$participant), healthy)
  
  df <- df %>% mutate(group = case_when(
    participant %in% healthy ~ "Healthy",
    participant %in% fibro   ~ "Fibro",
    TRUE                     ~ NA_character_
  ))
  
  list(trial_data = df,
       participants_with_7_blocks = participants_with_7,
       healthy_participants = healthy,
       fibro_participants   = fibro,
       non_learners         = non_learners)
}

result_healthy_fibro <- get_trial_data_healthy_fibro()

File: RL_Maggie/python/02_trial_etl_fibro.py
import pandas as pd
from pathlib import Path
from typing import List, Dict
from importlib import import_module

setup = import_module('RL_Maggie.python.00_setup')
LOCAL_DATA_DIR = setup.LOCAL_DATA_DIR


def get_trial_data_healthy_fibro(data_dir: Path = LOCAL_DATA_DIR) -> Dict[str, pd.DataFrame]:
    files = list(data_dir.glob('sub_*.csv'))
    df_list = []
    participants_with_7: List[str] = []

    for f in files:
        if f.stat().st_size == 0:
            print(f"Empty file: {f}")
            continue
        sub_df = pd.read_csv(f)
        sub_df = sub_df.dropna(subset=['trial_num'])
        if sub_df.empty:
            print(f"No trials in: {f}")
            continue

        sub_df = sub_df[['trial_num', 'participant', 'pair_type', 'choice_a', 'reward',
                         'correct', 'high_prob_image_file', 'low_prob_image_file']].copy()
        sub_df['block'] = (sub_df.index // 20) + 1

        cur_p = (sub_df.groupby('participant')['block'].max() == 7)
        participants_with_7.extend(sub_df.loc[cur_p[sub_df['participant']].index, 'participant'].unique())

        if sub_df['block'].max() == 7:
            sub_df = sub_df[sub_df['block'] != 1]
            sub_df['block'] -= 1
            sub_df['trial_num'] -= 20
        df_list.append(sub_df)

    if not df_list:
        return {'trial_data': pd.DataFrame(),
                'participants_with_7_blocks': [],
                'healthy_participants': [],
                'fibro_participants': [],
                'non_learners': []}
    df = pd.concat(df_list, ignore_index=True)
    df['correct'] = (df['correct'] == 'correct').astype(int)
    participants_with_7 = list(set(participants_with_7))

    if participants_with_7:
        non_learners = (df[df['participant'].isin(participants_with_7) & (df['block'] == 3)]
                        .groupby(['participant', 'pair_type'])['correct']
                        .mean().reset_index()
                        .pivot(index='participant', columns='pair_type', values='correct')
                        .reset_index())
        non_learners = non_learners[(non_learners.get('reversed', 0) < 0.7) |
                                    (non_learners.get('non-reversed', 0) < 0.7)]['participant'].tolist()
    else:
        non_learners = []

    p_nums = pd.to_numeric(df['participant'].str.replace('sub_', ''), errors='coerce')
    healthy = df.loc[p_nums.between(900, 1499), 'participant'].unique().tolist()
    fibro = [p for p in df['participant'].unique() if p not in healthy]

    df['group'] = df['participant'].apply(
        lambda p: 'Healthy' if p in healthy else ('Fibro' if p in fibro else None)
    )

    return {
        'trial_data': df,
        'participants_with_7_blocks': participants_with_7,
        'healthy_participants': healthy,
        'fibro_participants': fibro,
        'non_learners': non_learners
    }


if __name__ == '__main__':
    result_healthy_fibro = get_trial_data_healthy_fibro()
